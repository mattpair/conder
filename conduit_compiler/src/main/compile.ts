
import { Parse} from "./parse";
import { FunctionResolved, TypeResolved, Message, Field } from "./entity/resolved";
import { resolveDeps } from "./resolution/resolveTypes";
import { FileLocation } from "./util/filesystem";
import { Enum, EnumMember } from "./entity/basic";
import { resolveFunctions } from "./resolution/resolveFunction";

export type PartiallyCompiled = Record<string, {proto: string, data: FunctionResolved.File}>
export function compileFiles(files: Record<string, () => string>): PartiallyCompiled {
    const conduits: Parse.File[] = []
    for (const file in files) {
        if (file.endsWith(".cdt")) {
            conduits.push(Parse.extractAllFileEntities(files[file](), new FileLocation(file)))
        }
    }

    const r = resolveDeps(conduits)

    return resolveFunctionsAndProto(r)
} 

function resolveFunctionsAndProto(files: TypeResolved.File[]): PartiallyCompiled {
    const results: PartiallyCompiled = {}
    files.forEach(file => {

        const enums: Enum[] = []
        const messages: Message[] = []
        file.inFileScope.forEach(v => {
            switch(v.kind) {
                case "Enum":
                    enums.push(v)
                    break;
                case "Message":
                    messages.push(v)
                    break;
            }
        })
        let proto = ''
        if (enums.length + messages.length > 0) {
            proto = `
syntax="proto2";
${file.children.Import.map(d => `import "${d.dep.replace(".cdt", ".proto")}";`).join("\n")}
            
${enums.map(printEnum).join("\n\n")}
${messages.map(printMessage).join("\n\n")}
`
        }

        results[`${file.loc.fullname.replace(".cdt", ".proto")}`] = {proto, data: resolveFunctions(file)}
        
    })

    return results       
}

function printEnum(e: Enum): string {
    const mems = printMembers(e.children.EnumMember)
    return `
enum ${e.name} {
${mems}
}
    `
}

function printMembers(m: EnumMember[]): string {
    return m.map((e, index) => `\t${e.name} = ${index + 1};`).join("\n")
}

function printMessage(m: Message): string {
    const fields = printFields(m.children.Field)

    return `
message ${m.name} {
${fields}
}`
}

function printFields(fields: Field[]): string {
    return fields
    .map((f, index) => {
        const p = f.part.FieldType.differentiate()
        return `\t${f.isRequired ? 'required' : 'optional'} ${p.kind === "Primitive" ? p.val : p.name} ${f.name} = ${index + 1};`
    })
    .join("\n")
}