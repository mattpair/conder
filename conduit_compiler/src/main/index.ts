import * as fs from 'fs';
import {compileFiles} from "./compileToProto"

function conduitToProto(conduits: string[]): Promise<void[]>  {
    const toCompile: Record<string, () => string> = {}
    conduits.forEach(c => toCompile[c] = () => fs.readFileSync(`./conduit/${c}`, {encoding: "utf-8"}))
    const protos = compileFiles(toCompile)
    fs.mkdirSync(".proto")
    
    const writes = []
    for (const proto in protos) {
        console.log(`writing ${proto}`)
        writes.push(fs.promises.writeFile(`.proto/${proto}`, protos[proto]))
    }
    if (writes.length == 0) {
        console.warn("Did not find any message types in conduit/")
    }

    return Promise.all(writes)
}

function main() {
    let conduits: string[]
    try {
        conduits = fs.readdirSync("./conduit/")
    } catch(e) {
        console.error("Unable to find ./conduit/")
        return
    }

    if (conduits.length == 0) {
        console.warn("no files to compile")
    } else {
        conduitToProto(conduits)
        .then(() => console.log("done!"))
        .catch(() => console.log("failed."))
    }    
}

main()

