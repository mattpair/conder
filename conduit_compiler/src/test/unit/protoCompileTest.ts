import {compileFiles} from "../../main/compile"


function protoCompileTest(description: string, file: string) {
    test(description, () => {
        expect(compileFiles({"testFile.cdt": () => file})).toMatchSnapshot()
    })
} 

protoCompileTest("should convert simple conduit into proto2",
`
    struct Mystruct {
        double d,
        float f,
        int32 i32,
        int64 I,
        uint32 u32,
        uint64 u64,
        bool b,
        string s,
        bytes bs,
    }`
)

protoCompileTest("should convert multiple conduit into proto2", `
    struct m1 {
        double d,
        float f,
        int32 i32,
        int64 I,
        uint32 u32,
        uint64 u64,
    }
    struct m2{
        bool b,
        string s,
        bytes bs,
    }
    `)

protoCompileTest("should allow one struct to reference another", `
    struct m1 {
        double d,
        float f,
        int32 i32,
        int64 I,
        uint32 u32,
        uint64 u64,
    }
    struct m2{
        m1 myMessag,
    }
    `)

protoCompileTest("should allow the specification of optional fields", `
    struct Mystruct {
        optional double d,
        optional float f,
        optional int32 i32,
        optional int64 I,
        optional uint32 u32,
        optional uint64 u64,
        optional bool b,
        optional string s,
        optional bytes bs,
    }
    `)

protoCompileTest("should allow new line to end field", `
    struct Mystruct {
        optional double d
        float f
        optional int32 i32
        optional int64 I
        optional uint32 u32
        optional uint64 u64
        optional bool b
        optional string s
        optional bytes bs
    }
    `)

protoCompileTest("should allow enums",`
    enum PurchaseCategory {
        UNCATEGORIZED,
        GAS,
        DINING,
        TRAVEL,
        GROCERY,
        ENTERTAINMENT,
    }
    `)

protoCompileTest("should allow enums separated by whitespace",`
    enum PurchaseCategory {
        UNCATEGORIZED
        GAS
        DINING
        TRAVEL
        GROCERY
        ENTERTAINMENT
    }
    `)


protoCompileTest("should allow multiple enums separated by whitespace",`
    enum PurchaseCategory {
        UNCATEGORIZED
        GAS
        DINING
        TRAVEL
        GROCERY
        ENTERTAINMENT
    }
    
    enum OtherCategories {
        A
        B
        C
    }
    `)

protoCompileTest("defining a store", `
    struct stored {
        string s
    }

    myFirstStore = new Store<stored>
`)

protoCompileTest("getting all data from store", `
    struct stored {
        string s
    }

    secondStore = new Store<stored>

    function getAllData() stored[] {
        return all in secondStore
    }
`)