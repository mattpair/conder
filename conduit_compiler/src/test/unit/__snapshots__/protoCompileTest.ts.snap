// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow enums 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "PurchaseCategory" => Object {
          "children": Object {
            "EnumMember": Array [
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 22,
                  "endLineNumber": 2,
                  "startColNumber": 27,
                  "startLineNumber": 1,
                },
                "name": "UNCATEGORIZED",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 12,
                  "endLineNumber": 3,
                  "startColNumber": 22,
                  "startLineNumber": 2,
                },
                "name": "GAS",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 15,
                  "endLineNumber": 4,
                  "startColNumber": 12,
                  "startLineNumber": 3,
                },
                "name": "DINING",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 15,
                  "endLineNumber": 5,
                  "startColNumber": 15,
                  "startLineNumber": 4,
                },
                "name": "TRAVEL",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 16,
                  "endLineNumber": 6,
                  "startColNumber": 15,
                  "startLineNumber": 5,
                },
                "name": "GROCERY",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 22,
                  "endLineNumber": 7,
                  "startColNumber": 16,
                  "startLineNumber": 6,
                },
                "name": "ENTERTAINMENT",
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Enum",
          "loc": Object {
            "endColNumber": 27,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "PurchaseCategory",
        },
      },
      "size": 1,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should allow enums separated by whitespace 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "PurchaseCategory" => Object {
          "children": Object {
            "EnumMember": Array [
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 3,
                  "startColNumber": 27,
                  "startLineNumber": 1,
                },
                "name": "UNCATEGORIZED",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 4,
                  "startColNumber": 0,
                  "startLineNumber": 3,
                },
                "name": "GAS",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 5,
                  "startColNumber": 0,
                  "startLineNumber": 4,
                },
                "name": "DINING",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 6,
                  "startColNumber": 0,
                  "startLineNumber": 5,
                },
                "name": "TRAVEL",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 7,
                  "startColNumber": 0,
                  "startLineNumber": 6,
                },
                "name": "GROCERY",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 8,
                  "startColNumber": 0,
                  "startLineNumber": 7,
                },
                "name": "ENTERTAINMENT",
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Enum",
          "loc": Object {
            "endColNumber": 27,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "PurchaseCategory",
        },
      },
      "size": 1,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should allow multiple enums separated by whitespace 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "PurchaseCategory" => Object {
          "children": Object {
            "EnumMember": Array [
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 3,
                  "startColNumber": 27,
                  "startLineNumber": 1,
                },
                "name": "UNCATEGORIZED",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 4,
                  "startColNumber": 0,
                  "startLineNumber": 3,
                },
                "name": "GAS",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 5,
                  "startColNumber": 0,
                  "startLineNumber": 4,
                },
                "name": "DINING",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 6,
                  "startColNumber": 0,
                  "startLineNumber": 5,
                },
                "name": "TRAVEL",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 7,
                  "startColNumber": 0,
                  "startLineNumber": 6,
                },
                "name": "GROCERY",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 8,
                  "startColNumber": 0,
                  "startLineNumber": 7,
                },
                "name": "ENTERTAINMENT",
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Enum",
          "loc": Object {
            "endColNumber": 27,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "PurchaseCategory",
        },
        "OtherCategories" => Object {
          "children": Object {
            "EnumMember": Array [
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 12,
                  "startColNumber": 26,
                  "startLineNumber": 10,
                },
                "name": "A",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 13,
                  "startColNumber": 0,
                  "startLineNumber": 12,
                },
                "name": "B",
              },
              Object {
                "kind": "EnumMember",
                "loc": Object {
                  "endColNumber": 0,
                  "endLineNumber": 14,
                  "startColNumber": 0,
                  "startLineNumber": 13,
                },
                "name": "C",
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Enum",
          "loc": Object {
            "endColNumber": 26,
            "endLineNumber": 10,
            "startColNumber": 5,
            "startLineNumber": 8,
          },
          "name": "OtherCategories",
        },
      },
      "size": 2,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should allow new line to end field 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "MyMessage" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 2,
                  "startColNumber": 23,
                  "startLineNumber": 1,
                },
                "name": "d",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 3,
                  "startColNumber": 0,
                  "startLineNumber": 3,
                },
                "name": "f",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 4,
                  "startColNumber": 0,
                  "startLineNumber": 4,
                },
                "name": "i32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 5,
                  "startColNumber": 0,
                  "startLineNumber": 5,
                },
                "name": "I",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 6,
                  "startColNumber": 0,
                  "startLineNumber": 6,
                },
                "name": "u32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 7,
                  "startColNumber": 0,
                  "startLineNumber": 7,
                },
                "name": "u64",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 8,
                  "startColNumber": 0,
                  "startLineNumber": 8,
                },
                "name": "b",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 9,
                  "startColNumber": 0,
                  "startLineNumber": 9,
                },
                "name": "s",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 10,
                  "startColNumber": 0,
                  "startLineNumber": 10,
                },
                "name": "bs",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 23,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "MyMessage",
        },
      },
      "size": 1,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should allow one message to reference another 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "m1" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 2,
                  "startColNumber": 16,
                  "startLineNumber": 1,
                },
                "name": "d",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 3,
                  "startColNumber": 17,
                  "startLineNumber": 2,
                },
                "name": "f",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 4,
                  "startColNumber": 16,
                  "startLineNumber": 3,
                },
                "name": "i32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 5,
                  "startColNumber": 18,
                  "startLineNumber": 4,
                },
                "name": "I",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 6,
                  "startColNumber": 16,
                  "startLineNumber": 5,
                },
                "name": "u32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 7,
                  "startColNumber": 19,
                  "startLineNumber": 6,
                },
                "name": "u64",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 16,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "m1",
        },
        "m2" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 10,
                  "startColNumber": 15,
                  "startLineNumber": 9,
                },
                "name": "myMessag",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 15,
            "endLineNumber": 9,
            "startColNumber": 5,
            "startLineNumber": 8,
          },
          "name": "m2",
        },
      },
      "size": 2,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should allow the specification of optional fields 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "MyMessage" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 2,
                  "startColNumber": 23,
                  "startLineNumber": 1,
                },
                "name": "d",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 3,
                  "startColNumber": 26,
                  "startLineNumber": 2,
                },
                "name": "f",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 4,
                  "startColNumber": 25,
                  "startLineNumber": 3,
                },
                "name": "i32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 5,
                  "startColNumber": 27,
                  "startLineNumber": 4,
                },
                "name": "I",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 6,
                  "startColNumber": 25,
                  "startLineNumber": 5,
                },
                "name": "u32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 7,
                  "startColNumber": 28,
                  "startLineNumber": 6,
                },
                "name": "u64",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 8,
                  "startColNumber": 28,
                  "startLineNumber": 7,
                },
                "name": "b",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 9,
                  "startColNumber": 24,
                  "startLineNumber": 8,
                },
                "name": "s",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": false,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 17,
                  "endLineNumber": 10,
                  "startColNumber": 26,
                  "startLineNumber": 9,
                },
                "name": "bs",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 23,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "MyMessage",
        },
      },
      "size": 1,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should convert multiple conduit into proto2 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "m1" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 2,
                  "startColNumber": 16,
                  "startLineNumber": 1,
                },
                "name": "d",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 3,
                  "startColNumber": 17,
                  "startLineNumber": 2,
                },
                "name": "f",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 4,
                  "startColNumber": 16,
                  "startLineNumber": 3,
                },
                "name": "i32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 5,
                  "startColNumber": 18,
                  "startLineNumber": 4,
                },
                "name": "I",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 6,
                  "startColNumber": 16,
                  "startLineNumber": 5,
                },
                "name": "u32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 7,
                  "startColNumber": 19,
                  "startLineNumber": 6,
                },
                "name": "u64",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 16,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "m1",
        },
        "m2" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 10,
                  "startColNumber": 15,
                  "startLineNumber": 9,
                },
                "name": "b",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 11,
                  "startColNumber": 15,
                  "startLineNumber": 10,
                },
                "name": "s",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 12,
                  "startColNumber": 17,
                  "startLineNumber": 11,
                },
                "name": "bs",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 15,
            "endLineNumber": 9,
            "startColNumber": 5,
            "startLineNumber": 8,
          },
          "name": "m2",
        },
      },
      "size": 2,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;

exports[`should convert simple conduit into proto2 1`] = `
Object {
  "namespace": Object {
    "inScope": EntityMap {
      "map": Map {
        "MyMessage" => Object {
          "children": Object {
            "Field": Array [
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 2,
                  "startColNumber": 23,
                  "startLineNumber": 1,
                },
                "name": "d",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 3,
                  "startColNumber": 17,
                  "startLineNumber": 2,
                },
                "name": "f",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 4,
                  "startColNumber": 16,
                  "startLineNumber": 3,
                },
                "name": "i32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 5,
                  "startColNumber": 18,
                  "startLineNumber": 4,
                },
                "name": "I",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 6,
                  "startColNumber": 16,
                  "startLineNumber": 5,
                },
                "name": "u32",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 7,
                  "startColNumber": 19,
                  "startLineNumber": 6,
                },
                "name": "u64",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 8,
                  "startColNumber": 19,
                  "startLineNumber": 7,
                },
                "name": "b",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 9,
                  "startColNumber": 15,
                  "startLineNumber": 8,
                },
                "name": "s",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
              Object {
                "isRequired": true,
                "kind": "Field",
                "loc": Object {
                  "endColNumber": 8,
                  "endLineNumber": 10,
                  "startColNumber": 17,
                  "startLineNumber": 9,
                },
                "name": "bs",
                "part": Object {
                  "FieldType": Object {
                    "differentiate": [Function],
                    "kind": "FieldType",
                  },
                },
              },
            ],
          },
          "file": FileLocation {
            "dir": "",
            "fullname": "testFile.cdt",
            "name": "testFile.cdt",
          },
          "kind": "Message",
          "loc": Object {
            "endColNumber": 23,
            "endLineNumber": 1,
            "startColNumber": 0,
            "startLineNumber": 0,
          },
          "name": "MyMessage",
        },
      },
      "size": 1,
    },
    "name": "default",
  },
  "service": Object {
    "functions": Array [],
    "kind": "public",
  },
}
`;
